server/src/main/java/org/opensearch/gateway/AsyncShardBatchFetch.java

```diff
@@ -0,0 +1,341 @@
+        private final Map<Integer, ShardId> arrayToShardId;
+        private final Function<T, Map<ShardId, V>> shardsBatchDataGetter;
+        private final Supplier<V> emptyResponseBuilder;
+        private final Consumer<ShardId> failedShardHandler;
+        private final Function<V, Exception> getException;
+        private final Function<V, Boolean> isEmpty;
+        private final Logger logger;
+
+        public ShardBatchCache(
+            Logger logger,
+            String type,
+            Map<ShardId, ShardAttributes> shardAttributesMap,
+            String logKey,
+            Class<V> clazz,
+            BiFunction<DiscoveryNode, Map<ShardId, V>, T> responseGetter,
+            Function<T, Map<ShardId, V>> shardsBatchDataGetter,
+            Supplier<V> emptyResponseBuilder,
+            Consumer<ShardId> failedShardHandler,
+            Function<V, Exception> getResponseException,
+            Function<V, Boolean> isEmptyResponse
+        ) {
+            super(Loggers.getLogger(logger, "_" + logKey), type);
+            this.batchSize = shardAttributesMap.size();
+            this.getException = getResponseException;
+            this.isEmpty = isEmptyResponse;
+            cache = new HashMap<>();
+            shardIdToArray = new HashMap<>();
+            arrayToShardId = new HashMap<>();
+            fillShardIdKeys(shardAttributesMap.keySet());
+            this.shardResponseClass = clazz;
+            this.responseConstructor = responseGetter;
+            this.shardsBatchDataGetter = shardsBatchDataGetter;
+            this.emptyResponseBuilder = emptyResponseBuilder;
+            this.failedShardHandler = failedShardHandler;
+            this.logger = logger;
+        }
+
+        @Override
+        public Map<String, ? extends BaseNodeEntry> getCache() {
+            return cache;
+        }
+
+        @Override
+        public void deleteShard(ShardId shardId) {
+            if (shardIdToArray.containsKey(shardId)) {
+                Integer shardIdIndex = shardIdToArray.remove(shardId);
+                for (String nodeId : cache.keySet()) {
+                    cache.get(nodeId).clearShard(shardIdIndex);
+                }
+            }
+        }
+
+        @Override
+        public Map<DiscoveryNode, T> getCacheData(DiscoveryNodes nodes, Set<String> failedNodes) {
+            fillReverseIdMap();
+            return super.getCacheData(nodes, failedNodes);
+        }
+
+        /**
+         * Build a reverse map to get shardId from the array index, this will be used to construct the response which
+         * PrimaryShardBatchAllocator or ReplicaShardBatchAllocator are looking for.
+         */
+        private void fillReverseIdMap() {
+            arrayToShardId.clear();
+            for (Map.Entry<ShardId, Integer> indexMapping : shardIdToArray.entrySet()) {
+                arrayToShardId.putIfAbsent(indexMapping.getValue(), indexMapping.getKey());
+            }
+        }
```

Comment:
where is this ```arrayToShardId``` used?

Reply:
Removing it, directly using `shardIdToArray` entry set traversal.